app/assets/javascripts/discourse/app/routes/topic.js

```diff
@@ -183,10 +183,32 @@ const TopicRoute = DiscourseRoute.extend({
       }
 
       const topic = this.modelFor("topic");
+
       if (topic && currentPost) {
-        let postUrl = topic.get("url");
+        let postUrl;
+
         if (currentPost > 1) {
-          postUrl += "/" + currentPost;
+          postUrl = topic.urlForPostNumber(currentPost);
+        } else {
+          postUrl = topic.url;
+        }
+
+        if (this._router.currentRoute.queryParams) {
```

Comment:
Prior to this fix, we would drop the query params such as `username_filters`, `filters` and `replies_to_post_number`.